### PR TITLE
Initial read-only Snowflake Catalog implementation by @sfc-gh-mparmar

### PR DESCRIPTION
**Pull request by sfc-gh-dhuo**
Initial read-only Snowflake Catalog implementation by @sfc-gh-mparmar

### CODE REVIEW

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+    return sfTables.stream()
+        .map(
+            table ->
+                TableIdentifier.of(table.getDatabase(), table.getSchemaName(), table.getName()))
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
```
Comment (sfc-gh-dhuo) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:81]: We should probably throw UnsupportedOperationException or whatever is idiomatic within Iceberg instead of silently ignoring the ones we're not implementing.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    catalogProperties = properties;
+
+    String uri = properties.get(CatalogProperties.URI);
+    Preconditions.checkNotNull(uri, "JDBC connection URI is required");
+
+    if (name != null) {
+      this.catalogName = name;
+    }
+
+    LOG.debug("Connecting to JDBC database {}", properties.get(CatalogProperties.URI));
```
Comment (sfc-gh-dhuo) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:99]: Need to check on best practices for scrubbing sensitive secrets from log statements in case the URI is used with inline user/password strings.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeQueryFactory.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(SnowflakeQueryFactory.class);
+  private final JdbcClientPool connectionPool;
+
+  SnowflakeQueryFactory(JdbcClientPool conn) {
+    connectionPool = conn;
+  }
+
+  @Override
+  public List<SnowflakeSchema> listSchemas(Namespace namespace) {
+    String baseQuery = "show schemas";
```
Comment (sfc-gh-dhuo) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeQueryFactory.java:47]: Should extract string constants into static final strings.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/entities/SnowflakeTableMetadata.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      SnowflakeTableMetadata schema = new SnowflakeTableMetadata();
+      schema.jsonVal = rs.getString("LOCATION");
+      schema.metadataLocation = getMetadataLocationFromJson(schema.jsonVal);
+      schema.status = getStatusFromJson(schema.jsonVal);
+      return schema;
+    };
+  }
+
+  public static String getMetadataLocationFromJson(String json) {
+    return JsonUtil.parse(json, SnowflakeTableMetadata::getMetadataLocationFromJson);
```
Comment (sfc-gh-dhuo) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/entities/SnowflakeTableMetadata.java:54]: Instead of parsing separately each time we should probably just parse it initially and fetch fields from a JsonNode object consistently.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+package org.apache.iceberg.snowflake;
+
+import java.io.Closeable;
+import java.util.List;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.snowflake.entities.SnowflakeSchema;
+import org.apache.iceberg.snowflake.entities.SnowflakeTable;
+import org.apache.iceberg.snowflake.entities.SnowflakeTableMetadata;
+
```
Comment (sfc-gh-dhuo) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java:28]: Need javadoc comments on all the classes.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+public interface QueryFactory extends Closeable {
+  List<SnowflakeSchema> listSchemas(Namespace namespace);
```
Comment (sfc-gh-dhuo) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java:30]: If we're going to use this as an abstraction layer allowing us to easily swap out the bottom connectivity layer into Snowflake APIs, we should make the input argument also use Snowflake concepts, so that the business logic of translating an Iceberg Namespace into a Snowflake Database/Schema is common across implementations, while the way a list of tables is fetched given a db.schema can vary based on underlying implementations.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeQueryFactory.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+      // would be filtered to exclude tables from other databases.
+      else {
+        baseQuery += " in account";
+        final String finalQuery = baseQuery;
+        tables.addAll(
+            connectionPool.run(
+                conn -> run.query(conn, finalQuery, SnowflakeTable.createHandler())));
+      }
+
+      // In case of DB level namespace, filter the results to given namespace
```
Comment (sfc-gh-dhuo) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeQueryFactory.java:139]: It's unclear why this additional filter step is needed.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.io.Closeable;
+import java.util.List;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.snowflake.entities.SnowflakeSchema;
+import org.apache.iceberg.snowflake.entities.SnowflakeTable;
+import org.apache.iceberg.snowflake.entities.SnowflakeTableMetadata;
+
+public interface QueryFactory extends Closeable {
```
Comment (sfc-gh-dhuo) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java:29]: Naming-wise, I'd normally expected a QueryFactory to produce queries, rather than *execute* queries. Maybe better to name this class SnowflakeClient or something like that if the intent is to wrap JDBC and/or other API clients under a more constrained/usable interface.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeTableOperations.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  @Override
+  public void doRefresh() {
+    String location = null;
+    try {
+      LOG.debug("Getting metadata location for table {}", tableName());
+      location = getTableMetadataLocation();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```
Comment (sfc-gh-ppaulus) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeTableOperations.java:66]: Why are we doing this?

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeTableOperations.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    } catch (SQLException e) {
+      // SQL exception happened when getting table from catalog
+      throw new UncheckedSQLException(
+          e, "Failed to get table %s from catalog %s", tableIdentifier, catalogName);
+    }
+
+    if (location == null || location.isEmpty()) {
+      if (currentMetadataLocation() != null) {
+        throw new NoSuchTableException(
+            "Failed to load table %s from catalog %s: dropped by another process",
```
Comment (sfc-gh-ppaulus) [inline snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeTableOperations.java:77]: Do we know for sure that it was dropped by another process? Maybe we should just say "table could not be found".

File: snowflake/src/test/java/org/apache/iceberg/snowflake/SnowflakeCatalogTest.java [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+        if (line.trim().length() > 0) {
+          st.execute(line);
+        }
+      }
+    }
+  }
+
+  @Test
+  public void testListNamespace() {
```
Comment (sfc-gh-ppaulus) [inline snowflake/src/test/java/org/apache/iceberg/snowflake/SnowflakeCatalogTest.java:121]: We're testing the golden path scenarios here, but I don't see us testing error and edge cases, like an empty namespace, or a non-existent database or schema, or too many namespace levels. Can we add tests that cover our error scenarios too?

File: snowflake/src/test/java/org/apache/iceberg/snowflake/TestQueryFactory.java [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+            connectionPool.run(
+                conn ->
+                    run.query(
+                        conn,
+                        finalQuery,
+                        SnowflakeSchema.createHandler(),
+                        namespace.level(SnowflakeResources.NAMESPACE_DB_LEVEL - 1)));
+      }
+    } catch (SQLException | InterruptedException e) {
+      throw new RuntimeException(e);
```
Comment (sfc-gh-ppaulus) [inline snowflake/src/test/java/org/apache/iceberg/snowflake/TestQueryFactory.java:64]: Why are we throwing RuntimeExceptions?

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeQueryFactory.java (hunk 7 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeQueryFactory.java:139]: Ah makes sense, thanks for the context. I can clean it up.

File: snowflake/src/test/java/org/apache/iceberg/snowflake/SnowflakeCatalogTest.java [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+      Path path =
+          Paths.get(
+              Objects.requireNonNull(
+                      SnowflakeCatalogTest.class.getClassLoader().getResource(TEST_SETUP_SCRIPT))
+                  .toURI());
+      importSQLScript(con, new FileInputStream(path.toFile()));
+
+      catalog = new SnowflakeCatalog();
+      JdbcClientPool connectionPool = new JdbcClientPool(uri, properties);
+      catalog.setQueryFactory(new TestQueryFactory(connectionPool));
```
Comment (sfc-gh-dhuo) [inline snowflake/src/test/java/org/apache/iceberg/snowflake/SnowflakeCatalogTest.java:93]: Since the underlying JDBC here won't be exercising a real Snowflake data model anyways, IMO it's better to keep this unittest contained to the responsibilities of SnowflakeCatalog itself, which is namely simply to translate between our own in-memory data model (entities/) into Iceberg idioms. Then we don't need a fake underlying database or a fake query factory.

We could also try to unittest the query factory (which we should rename to JdbcSnowflakeClient) by fake or mock JdbcClientPools and just verify the queries generated rather than using a fake sqlite database.

File: snowflake/src/test/java/org/apache/iceberg/snowflake/TestQueryFactory.java [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+import org.apache.commons.dbutils.QueryRunner;
+import org.apache.iceberg.catalog.Namespace;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.jdbc.JdbcClientPool;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.snowflake.entities.SnowflakeSchema;
+import org.apache.iceberg.snowflake.entities.SnowflakeTable;
+import org.apache.iceberg.snowflake.entities.SnowflakeTableMetadata;
+
+public class TestQueryFactory implements QueryFactory {
```
Comment (sfc-gh-dhuo) [inline snowflake/src/test/java/org/apache/iceberg/snowflake/TestQueryFactory.java:32]: Generally the naming convention I see elsewhere in Iceberg is TestFoo is a file containing test cases for Foo.

I'd rename these things to "FakeFoo" - typically tests use Mocks (for micromanaged interactions) or Fakes (for in-memory "fake" implementations) as the usual naming convention.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {}
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    Preconditions.checkArgument(
+        namespace.length() <= SnowflakeResources.MAX_NAMESPACE_DEPTH,
+        "Snowflake doesn't supports more than 2 levels of namespace");
+    List<Namespace> namespaceList = Lists.newArrayList();
```
[the remark this answers is not in the document]
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:129]: IIRC it started mostly as a pre-Java9 convention where you'd have to duplicate the generic type like "new ArrayList<Namespace>()", and also to avoid having to import both java.util.List and java.util.ArrayList; in this case I guess we can just follow whatever the rest of the Iceberg project is doing for consistency.

    [dhuo@SDP_DevVM-dhuo ~/Snowflake-Labs/iceberg]$ grep -I -R Lists.newArrayList * | wc -l
    3050
    [dhuo@SDP_DevVM-dhuo ~/Snowflake-Labs/iceberg]$ grep -I -R "new ArrayList" * | wc -l
    22
    [dhuo@SDP_DevVM-dhuo ~/Snowflake-Labs/iceberg]$ grep -I -R "new ArrayList<>" * | wc -l
    10

Apparently they also helped make it explicit by encoding it in their checkstyle rules, so that's reassuring that we'll have some compiler guidance on project style-preferences:

https://github.com/apache/iceberg/blob/86156757653112b1a89d7e7f08903ef746fd918b/.baseline/checkstyle/checkstyle.xml#L60

    <module name="RegexpSingleline">
        <property name="format" value="new ArrayList&lt;&gt;\(.*\)"/>
        <property name="message" value="Prefer using Lists.newArrayList() instead."/>
    </module>

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+  public void setQueryFactory(QueryFactory factory) {
+    queryFactory = factory;
+  }
+
+  @Override
+  public List<TableIdentifier> listTables(Namespace namespace) {
+    Preconditions.checkArgument(
+        namespace.length() <= SnowflakeResources.MAX_NAMESPACE_DEPTH,
+        "Snowflake doesn't supports more than 2 levels of namespace");
```
[the remark this answers is not in the document]
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:77]: Done

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+    fileIO = CatalogUtil.loadFileIO(fileIOImpl, catalogProperties, conf);
+  }
+
+  @Override
+  public void close() {
+    queryFactory.close();
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {}
```
[the remark this answers is not in the document]
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:128]: Fixed, added throw new UnsupportedOperationException (which is in-line with the base catalog impl) here and other unimplemented methods.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/entities/SnowflakeTableMetadata.java (hunk 4 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/entities/SnowflakeTableMetadata.java]: Fixed

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java (hunk 1 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:89]: Done

File: snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java (hunk 5 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java]: Added for classes that can benefit from explanation

File: snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java (hunk 8 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java]: Done

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java (hunk 2 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java]: Removed

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    Preconditions.checkArgument(
+        namespace.length() <= SnowflakeResources.MAX_NAMESPACE_DEPTH,
+        "Snowflake doesn't supports more than 2 levels of namespace");
```
[the remark this answers is not in the document]
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:135]: Done

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+    List<Namespace> namespaceList = Lists.newArrayList();
+    try {
+      List<SnowflakeSchema> sfSchemas = queryFactory.listSchemas(namespace);
+      namespaceList =
+          sfSchemas.stream()
+              .map(schema -> Namespace.of(schema.getDatabase(), schema.getName()))
+              .collect(Collectors.toList());
+    } catch (UncheckedSQLException | UncheckedInterruptedException ex) {
+      LOG.error("{}", ex.getMessage(), ex);
```
[the remark this answers is not in the document]
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:144]: Removed this try/catch; at this level in the code, unchecked exceptions should probably just propagate all the way up.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    if (namespace.length() == SnowflakeResources.MAX_NAMESPACE_DEPTH) {
```
[the remark this answers is not in the document]
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:147]: Looks like the logic here is actually for an edge case where we're trying to list schemas but already specified the exact schema we want. As in:

listNamespaces("foo_database.bar_schema");

and the logic in the JDBC layer currently just lists all schemas under "foo_database" and then this filter makes it so:

1. If the schema doesn't exist, we return empty list
2. If it does exist, we filter it down to a size-1 list "equal" to the input

Thinking about it more though, I'm not sure listing schemas in a schema should behave like an existence-check.

When listing namespaces under the "database" level, the 1-level database namespace isn't returned as one of the namespaces, even though the 1-level namespace is allowed in Spark's "in" clause:

    scala> spark.sql("show namespaces in iot_data").show(false);
    +---------------------------+
    |namespace                  |
    +---------------------------+
    |IOT_DATA.ICE_SCHEMA        |
    |IOT_DATA.INFORMATION_SCHEMA|
    |IOT_DATA.PUBLIC            |
    |IOT_DATA.TPCDS10TB         |
    +---------------------------+

So that implies to me that listing "in" a two-level namespace is supposed to only return 3-level namespaces, which aren't supported, and thus we should just throw IllegalArgumentException in that case.

Fixed the Precondition to require level <= MAX - 1

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+
+  @Override
+  public boolean removeProperties(Namespace namespace, Set<String> properties) {
+    return false;
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
+    Preconditions.checkArgument(
+        tableIdentifier.namespace().length() <= SnowflakeResources.MAX_NAMESPACE_DEPTH,
```
[the remark this answers is not in the document]
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:192]: Done

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeTableOperations.java (hunk 9 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeTableOperations.java:66]: Looks like this follows convention of JdbcCatalog: https://github.com/apache/iceberg/blob/0694269cd20a3625d5587d544bec621e9bec305a/core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java#L114

I'm also a bit fuzzy on the actual scenarios where the thread's interrupt bit becomes important when we're going to rethrow an unchecked exception anyways, but this stackoverflow thread seems to provide a good overview. I think the problem comes from the fact that ClientPoolImpl declares throwing the InterruptedException: https://github.com/apache/iceberg/blob/0694269cd20a3625d5587d544bec621e9bec305a/core/src/main/java/org/apache/iceberg/ClientPoolImpl.java#L55

Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeTableOperations.java:66]: Actually, it looks like this is self-inflicted, since we already wrap all the exceptions in the underlying JDBC client wrapper. So this TableOperations class shouldn't need to duplicate the exception wrapping. I'll clean it up.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeTableOperations.java (hunk 10 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeTableOperations.java:77]: Looks like all this is vestigial; this can't happen in the current code because the metadata object itself is null for missing tables and we throw NoSuchTableException below for that case. Changed the location check to a precondition and removed this whole conditional branch.

File: snowflake/src/test/java/org/apache/iceberg/snowflake/SnowflakeCatalogTest.java (hunk 13 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/test/java/org/apache/iceberg/snowflake/SnowflakeCatalogTest.java]: Done

File: snowflake/src/test/java/org/apache/iceberg/snowflake/TestQueryFactory.java (hunk 12 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/test/java/org/apache/iceberg/snowflake/TestQueryFactory.java]: Rewrote this whole class as an in-memory fake instead of sqllite fake, removed these exceptions

File: snowflake/src/test/java/org/apache/iceberg/snowflake/TestFileIO.java [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+
+public class TestFileIO implements FileIO {
+  @Override
+  public InputFile newInputFile(String path) {
+    try {
+      Path filePath = Paths.get(getClass().getClassLoader().getResource(path).toURI());
+      byte[] data = Files.readAllBytes(filePath);
+      return (InputFile) new InMemoryInputFile(filePath.toString(), data);
```
[the remark this answers is not in the document]
Comment (sfc-gh-dhuo) [reply snowflake/src/test/java/org/apache/iceberg/snowflake/TestFileIO.java]: Done

File: snowflake/src/test/java/org/apache/iceberg/snowflake/TestQueryFactory.java (hunk 14 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/test/java/org/apache/iceberg/snowflake/TestQueryFactory.java]: Done

File: snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java (hunk 6 above)
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/QueryFactory.java]: Since there's actually a fair amount of JDBC-specific logic in translating the namespace into the right predicate or identifier in the SQL statement, I'll wait until we add another lower-level client impl before extracting the namespace-conversion out, and for now just focus on refacting within JdbcSnowflakeClient to reduce duplication of boilerplate.

**Comment by sfc-gh-dhuo** [comment]
@sfc-gh-ppaulus  I'll still need to work on adding additional test cases, but all the major refactor suggestions from the first round of reviews is done, PTAL

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (sfc-gh-ppaulus) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java]: Agree, thanks.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (sfc-gh-ppaulus) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java]: Ok, thanks.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeQueryFactory.java [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+        tables.removeIf(
+            table ->
+                !table
+                    .getDatabase()
+                    .equalsIgnoreCase(namespace.level(SnowflakeResources.NAMESPACE_DB_LEVEL - 1)));
+      }
+
+    } catch (SQLException e) {
+      LOG.error("Failed to list tables for namespace {}", namespace, e);
+      throw new UncheckedSQLException(
```
[the remark this answers is not in the document]
Comment (sfc-gh-ppaulus) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeQueryFactory.java]: Ugh, ok. Let's make sure that habit doesn't creep into our service code. :)

**Comment by sfc-gh-ppaulus** [comment]
> @sfc-gh-ppaulus I'll still need to work on adding additional test cases, but all the major refactor suggestions from the first round of reviews is done, PTAL

The changes look reasonable. Does this include the changes to support key files?

**Comment by sfc-gh-dhuo** [comment]
> > @sfc-gh-ppaulus I'll still need to work on adding additional test cases, but all the major refactor suggestions from the first round of reviews is done, PTAL
> 
> The changes look reasonable. Does this include the changes to support key files?

Not yet, planning to take some time to figure out how to test keyfile auth today.

**Comment by sfc-gh-dhuo** [comment]
@sfc-gh-ppaulus Good news - though I had assumed the worst for keyfile support due to how other libraries implemented it, I managed to confirm that the current code already passes through sufficient properties to the JDBC driver such that no new code is needed at all to fully support all the keyfile passthrough options. I successfully tested end-to-end with:

`--conf spark.sql.catalog.snowlog.jdbc.user=dhuo --conf spark.sql.catalog.snowlog.jdbc.private_key_file=/home/dhuo/auth-keys/rsa_key.p8`

There are some quirks to how password-encrypted keyfiles need security-provider setup, but that part is non-specific to this Iceberg library, and is the same for all direct users of the Snowflake JDBC driver as well.

So the only additional changes will be expanding unittests and doing some path-translation for Azure/GCS.

File: snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java [hunk 24]
```diff
@@ -0,0 +1,220 @@
...
+      String uri = properties.get(CatalogProperties.URI);
+      Preconditions.checkNotNull(uri, "JDBC connection URI is required");
+
+      try {
+        // We'll ensure the expected JDBC driver implementation class is initialized through
+        // reflection
+        // regardless of which classloader ends up using this JdbcSnowflakeClient, but we'll only
+        // warn if the expected driver fails to load, since users may use repackaged or custom
+        // JDBC drivers for Snowflake communcation.
+        Class.forName(JdbcSnowflakeClient.EXPECTED_JDBC_IMPL);
```
[the remark this answers is not in the document]
Comment (sfc-gh-dhuo) [reply snowflake/src/main/java/org/apache/iceberg/snowflake/SnowflakeCatalog.java:116]: Good question - hadn't had a chance to test it yet, but I've tested it successfully now on EMR, as well as the error case.

    # Works
    spark-shell --packages net.snowflake:snowflake-jdbc:3.13.22 --jars iceberg-spark-runtime-3.2_2.12-instance-initialization-forname.jar --conf ...

    # Error
    spark-shell --packages net.snowflake:snowflake-jdbc:3.13.22 --jars iceberg-spark-runtime-3.2_2.12-no-forname.jar --conf ...

I also tried with it in the static block like yours and it happened to work on my EMR cluster. However, as far as I understand it, the reason it's a gamble to put a `Class.forName` call in a static initalization block especially when using Spark is that there's a lot of fancy dynamic jar-downloading and classloading stuff, such that I wouldn't count on the classpath resources necessarily being complete at the time the static initializer block is called.

One possibility is if you tested with the Snowflake JDBC Driver being a `--jars` argument instead of `--packages` argument since I tested it in `--packages`?

File: snowflake/src/test/java/org/apache/iceberg/snowflake/FakeSnowflakeClient.java [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.jdbc.UncheckedSQLException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.snowflake.entities.SnowflakeSchema;
+import org.apache.iceberg.snowflake.entities.SnowflakeTable;
+import org.apache.iceberg.snowflake.entities.SnowflakeTableMetadata;
+
+public class FakeSnowflakeClient implements SnowflakeClient {
```
[the remark this answers is not in the document]
Comment (sfc-gh-dhuo) [reply snowflake/src/test/java/org/apache/iceberg/snowflake/FakeSnowflakeClient.java:33]: Right, my motivation for the refactor was specifically to make SnowflakeCatalogTest agnostic of the client implementation - that's the goal to make the unittest well-contained. We could always also add a unittest that's more of an integration test if we have a good way to mock out the lower JDBC communication, but ultimately if we supply a fake SnowflakeClient anyways that doesn't call into JdbcSnowflakeClient, that client should avoid using JDBC under the hood.

If we can inject something into a real JdbcSnowflakeClient that achieves what we want then we can inject that real JdbcSnowflakeClient into a real SnowflakeCatalog for a more end-to-end test.

Otherwise, for single-class unittests we can keep them clean and self-contained.

File: snowflake/src/test/java/org/apache/iceberg/snowflake/SnowflakeCatalogTest.java (hunk 11 above)
Comment (dennishuo) [reply snowflake/src/test/java/org/apache/iceberg/snowflake/SnowflakeCatalogTest.java:125]: Done
